crates/red_knot/src/module.rs

```diff
@@ -768,19 +782,205 @@ impl PackageKind {
     }
 }
 
+#[derive(Debug)]
+struct TypeshedVersions(FxHashMap<ModuleName, PyVersionRange>);
+
+#[allow(unused)]
+impl TypeshedVersions {
+    fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    fn contains_module(&self, module_name: impl Into<ModuleName>) -> bool {
+        self.0.contains_key(&module_name.into())
+    }
+
+    fn module_exists_on_version(
+        &self,
+        module: impl Into<ModuleName>,
+        version: impl Into<PyVersion>,
+    ) -> bool {
+        let version = version.into();
+        let mut module = Some(module.into());
+        while let Some(module_to_try) = module {
+            if let Some(range) = self.0.get(&module_to_try) {
+                return range.contains(version);
+            }
+            module = module_to_try.parent();
+        }
```

Comment:
The most common situation is that there's an exact match in `VERSIONS` -- you're looking to find out the versions `pathlib` exists on, and `VERSIONS` has an entry called `pathlib` in it.

The second most common situation is probably that you're looking up a submodule and the submodule itself doesn't exist in `VERSIONS` but its parent package does -- e.g. you're looking up `email.utils`, which doesn't exist in `VERSIONS`, but `email` does. But I'm not sure we can optimize that, because if `email.utils` _also_ exists in `VERSIONS` as _well_ as `email`, we _have_ to give that entry precedence over the entry for `email`.